python/ray/experimental/channel/nccl_group.py

```diff
@@ -0,0 +1,160 @@
+from types import ModuleType
+from typing import TYPE_CHECKING, Dict, Optional
+
+import ray
+
+if TYPE_CHECKING:
+    import cupy as cp
+    import torch
+
+
+class _NcclGroup:
+    """
+    Represents an actor's NCCL communicator.
+
+    This class is not thread-safe.
+    """
+
+    def __init__(
+        self,
+        world_size: int,
+        comm_id: int,
+        rank: Optional[int],
+        actor_ids_to_ranks: Dict[ray.ActorID, int],
+        cuda_stream: Optional[int],
+    ):
+        """
+        Initialize a NCCL communicator that can be used to communicate p2p with
+        other GPU actors.
+
+        This method blocks until the same call has been made on all other
+        actors in the group, with the same arguments for world_size and
+        comm_id.
+
+        Args:
+            world_size: The number of participating actors/devices.
+            comm_id: A unique communicator ID returned by
+                cupy.cuda.nccl.get_unique_id().
+            rank: The rank of this actor. If None, then the caller is not a
+                participant of the NCCL group.
+            actor_ids_to_ranks: A map from actor ID to its rank in the group.
+            cuda_stream: A raw CUDA stream to dispatch NCCL ops to. If rank is
+                specified, then this must be specified too.
+        """
+        self._rank: Optional[int] = rank
+        self.nccl_util: Optional[ModuleType] = None
+        self._actor_ids_to_ranks = actor_ids_to_ranks
+
+        if rank is not None:
+            assert ray.get_gpu_ids(), "NCCL actor has no GPUs assigned"
+            assert cuda_stream is not None, "NCCL actor must specify cuda_stream"
```

Comment:
We should also check that if cuda_stream is not None, then rank is also not None. Currently, the following would not be caught:
rank = None
cuda_stream = `some instance`